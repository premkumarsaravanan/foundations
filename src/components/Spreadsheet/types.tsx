import * as React from 'react'
import ReactDataSheet from 'react-datasheet'

export class MyReactDataSheet extends ReactDataSheet<Cell, string> {}

/** Cell contain predefined value
 * https://github.com/nadbm/react-datasheet/blob/master/types/react-datasheet.d.ts
 * plus some properties below
 */
export interface Cell extends ReactDataSheet.Cell<Cell, string> {
  /** The value of the cell, always a string */
  value: string
  /** The validate function, receive Cell as param, must return boolean */
  validate?: (cell: Cell) => boolean
  /** Additional className for styling cell */
  className?: string
  style?: React.CSSProperties
  CustomComponent?: React.FC<{
    data: Cell[][]
    cellRenderProps: ReactDataSheet.CellRendererProps<Cell>
    setData: React.Dispatch<Cell[][]>
    setSelected: React.Dispatch<SelectedMatrix>
  }>
}

export interface CellWithValidate extends Cell {
  validate: (cell: Cell) => boolean
}

export interface DoubleClickPayLoad {
  row: number
  col: number
  maxRowIndex: number
  maxColIndex: number
  isReadOnly?: boolean
}

export interface SpreadsheetProps {
  data: Cell[][]
  description?: React.ReactNode
  hasUploadButton?: boolean
  hasDownloadButton?: boolean
  hasAddButton?: boolean
  /** This will run before data is set to table, must return data with validate function */
  validateUpload?: (data: Cell[][]) => CellWithValidate[][]
}

export interface SelectedMatrix {
  start: ReactDataSheet.Location
  end: ReactDataSheet.Location
}
