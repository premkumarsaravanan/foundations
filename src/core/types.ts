export interface LoginSession {
  accessToken: string
  accessTokenExpiry: number
  idToken: string
  idTokenExpiry: number
  refreshToken: string
}

export interface LoginParams {
  userName: string
  password: string
}

export interface ChangePasswordParams {
  newPassword: string
  userName: string
  password: string
}

export interface RefreshParams {
  userName: string
  refreshToken: string
}

export interface ConfirmPasswordParams {
  userName: string
  newPassword: string
  verificationCode: string
}

export interface ConfirmRegistrationParams {
  userName: string
  verificationCode: string
}

export interface ResetPasswordParams {
  userName: string
}
