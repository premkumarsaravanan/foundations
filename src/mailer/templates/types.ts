export interface ForgottenPasswordTemplateParams {
  verificationCode: string
  userName: string
}

export interface ConfirmPasswordTemplateParams {
  userName: string
}
