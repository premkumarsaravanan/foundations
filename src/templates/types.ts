export interface ForgottenPasswordTemplateParams {
  verificationCode: string
  userName: string
}
