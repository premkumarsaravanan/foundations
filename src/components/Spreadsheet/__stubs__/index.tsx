import * as React from 'react'
import { Cell, SelectedMatrix } from '../types'
import ReactDataSheet from 'react-datasheet'

export const data: Cell[][] = [
  [
    { readOnly: true, value: 'Office Name' },
    { readOnly: true, value: 'Building Name' },
    { readOnly: true, value: 'Building No.' },
    { readOnly: true, value: 'Address 1' },
    { readOnly: true, value: 'Address 2' },
    { readOnly: true, value: 'Address 3' },
    { readOnly: true, value: 'Address 4' },
    { readOnly: true, value: 'Post Code' },
    { readOnly: true, value: 'Telephone' },
    { readOnly: true, value: 'Fax' },
    { readOnly: true, value: 'Email' }
  ],
  [
    { value: 'London' },
    { value: 'The White House' },
    { value: '15' },
    { value: 'London 1' },
    { value: '' },
    { value: 'Londom 3' },
    { value: '' },
    { value: 'EC12NH' },
    { value: '0845 0000' },
    { value: '' },
    { value: 'row1@gmail.com' }
  ],
  [
    { value: 'London2' },
    { value: 'The Black House' },
    { value: '11' },
    { value: '' },
    { value: '' },
    { value: 'Adress 3' },
    { value: '' },
    { value: 'EC12NH' },
    { value: '087 471 929' },
    { value: '' },
    { value: 'row2@gmail.com' }
  ],
  [
    { value: 'New York' },
    { value: 'Building A' },
    { value: '11' },
    { value: '' },
    { value: '' },
    { value: 'City Z' },
    { value: '' },
    { value: 'AL7187' },
    { value: '017 7162 9121' },
    { value: '' },
    { value: 'row3@gmail.com' }
  ]
]

export const cellRenderProps: ReactDataSheet.CellRendererProps<Cell> = {
  row: 3,
  col: 10,
  cell: { value: 'row3@gmail.com' },
  selected: false,
  editing: false,
  updated: false,
  attributesRenderer: jest.fn() as ReactDataSheet.AttributesRenderer<Cell>,
  className: 'cell',
  style: { background: 'red' },
  onMouseDown: jest.fn(),
  onMouseOver: jest.fn(),
  onDoubleClick: jest.fn(),
  onContextMenu: jest.fn(),
  children: <div>hi</div>
}

export const setData: React.Dispatch<Cell[][]> = jest.fn()

export const setSelected: React.Dispatch<SelectedMatrix> = jest.fn()

export const selectedMatrix = {
  start: { i: 0, j: 1 },
  end: { i: 2, j: 3 }
}
