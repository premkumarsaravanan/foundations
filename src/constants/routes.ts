const Routes = {
  CLIENT: '/client/apps',
  CLIENT_PAGINATE: '/client/apps/:page',
  MY_APPS: '/client/installed',
  MY_APPS_PAGINATE: '/client/installed/:page',
  DEVELOPER: '/developer',
  DEVELOPER_MY_APPS: '/developer/apps',
  DEVELOPER_SWAGGER: '/developer/swagger',
  DEVELOPER_ELEMENTS: '/developer/elements',
  DEVELOPER_MY_APPS_PAGINATE: '/developer/apps/:page',
  DEVELOPER_API_DOCS: '/developer/api-docs',
  DEVELOPER_DESKTOP_DOCS: '/developer/desktop-docs',
  DEVELOPER_WEB_COMPONENTS: '/developer/web-components',
  DEVELOPER_ANALYTICS: '/developer/analytics',
  SETTINGS: '/developer/settings',
  SUBMIT_APP: '/developer/submit-app',
  REGISTER: '/register',
  ADMIN: '/admin',
  ADMIN_APPROVALS: '/admin/approvals',
  ADMIN_APPROVALS_PAGINATE: '/admin/approvals/:page',
  ADMIN_LOGIN: '/admin/login',
  CLIENT_LOGIN: '/client/login',
  DEVELOPER_LOGIN: '/developer/login',
  FOUR_O_FOUR: '/404'
}

export default Routes
