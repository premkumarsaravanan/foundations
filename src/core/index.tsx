import '../styles/index.scss'
import * as React from 'react'
import { render } from 'react-dom'
import Router from './router'
import { Provider } from 'react-redux'
import store from './store'
import Toast from '../components/ui/toast'
import { PortalProvider } from '@reapit/elements'

const rootElement = document.querySelector('#root') as Element

const App = () => (
  <Provider store={store.reduxStore}>
    <PortalProvider>
      <Router />
    </PortalProvider>
    <Toast />
  </Provider>
)

if (rootElement) {
  render(<App />, rootElement)
}

export default App
