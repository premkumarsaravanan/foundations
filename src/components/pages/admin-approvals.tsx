import * as React from 'react'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { AdminApprovalsState } from '@/reducers/admin-approvals'
import { Loader, Pagination } from '@reapit/elements'
import ErrorBoundary from '@/components/hocs/error-boundary'
import { withRouter, RouteComponentProps } from 'react-router'
import { oc } from 'ts-optchain'
import routes from '@/constants/routes'
import { AppDetailState } from '@/reducers/app-detail'
import bulma from '@/styles/vendor/bulma'
import { REVISIONS_PER_PAGE } from '@/constants/paginator'
import { revisionDetailRequestData, RevisionDetailRequestParams } from '@/actions/revision-detail'
import AdminApprovalModal from '../ui/admin-approval-modal'
import { appDetailRequestData } from '@/actions/app-detail'
import { RevisionDetailState } from '@/reducers/revision-detail'
import Info from '../pages/info'

export interface AdminApprovalsMappedActions {
  fetchRevisionDetail: (params: RevisionDetailRequestParams) => void
  fetchAppDetail: (id: string) => void
}

export interface AdminApprovalsMappedProps {
  approvalsState: AdminApprovalsState
  appDetail: AppDetailState
  revisionDetail: RevisionDetailState
}

export type AdminApprovalsProps = AdminApprovalsMappedActions &
  AdminApprovalsMappedProps &
  RouteComponentProps<{ page?: any }>

export const AdminApprovals: React.FunctionComponent<AdminApprovalsProps> = ({
  approvalsState,
  match,
  history,
  fetchRevisionDetail,
  fetchAppDetail,
  appDetail,
  revisionDetail
}) => {
  const pageNumber = match.params && !isNaN(match.params.page) ? Number(match.params.page) : 1
  const unfetched = !approvalsState.adminApprovalsData
  const loading = approvalsState.loading
  const list = oc<AdminApprovalsState>(approvalsState).adminApprovalsData.data.data([])
  const { totalCount, pageSize } = oc<AdminApprovalsState>(approvalsState).adminApprovalsData.data({})
  const [isModalOpen, setIsModalOpen] = React.useState(false)
  const pageNumberInState = oc<AdminApprovalsState>(approvalsState).adminApprovalsData.data.pageNumber(1)

  if (unfetched && loading) {
    return <Loader />
  }
  return (
    <ErrorBoundary>
      <div className={`${bulma.container} ${bulma.isRelative} py-8`} data-test="revision-list-container">
        {loading && (
          <div className="pin absolute flex items-center justify-center">
            <Loader />
          </div>
        )}
        {!loading && !list.length ? (
          <Info infoType="ADMIN_APPROVALS_EMPTY" />
        ) : (
          <table className={`${bulma.table} ${bulma.isFullwidth}`} data-test="revision-list">
            <thead>
              <tr>
                <th>#</th>
                <th>AppId</th>
                <th>Type</th>
                <th>Description</th>
                <th />
              </tr>
            </thead>
            <tbody>
              {list.map((revision, index) => (
                <tr key={revision.appRevisionId} data-app-id={revision.appId}>
                  <th>{(pageNumberInState - 1) * REVISIONS_PER_PAGE + index + 1}</th>
                  <th>{revision.appId}</th>
                  <th>{revision.type}</th>
                  <th>{revision.description}</th>
                  <th>
                    <button
                      data-test={`view-details-button_${revision.appId}`}
                      className={`${bulma.button} ${bulma.isPrimary}`}
                      onClick={() => {
                        const { appId, appRevisionId } = revision
                        if (appRevisionId && appId) {
                          const currentRevisionId = oc<RevisionDetailState>(revisionDetail).revisionDetailData.data.id(
                            undefined
                          )
                          const currentAppId = oc<AppDetailState>(appDetail).appDetailData.data.id(undefined)
                          if (currentRevisionId !== appRevisionId) {
                            fetchRevisionDetail({ appId, appRevisionId })
                          }
                          if (currentAppId !== appId) {
                            fetchAppDetail(appId)
                          }
                          setIsModalOpen(true)
                        }
                      }}
                    >
                      View details
                    </button>
                  </th>
                </tr>
              ))}
            </tbody>
          </table>
        )}
        <Pagination
          onChange={page => history.push(`${routes.ADMIN_APPROVALS}/${page}`)}
          totalCount={totalCount}
          pageSize={pageSize}
          pageNumber={pageNumber}
        />
      </div>
      <AdminApprovalModal visible={isModalOpen} afterClose={() => setIsModalOpen(false)} />
    </ErrorBoundary>
  )
}

const mapStateToProps = (state: ReduxState): AdminApprovalsMappedProps => ({
  approvalsState: state.adminApprovals,
  appDetail: state.appDetail,
  revisionDetail: state.revisionDetail
})

const mapDispatchToProps = (dispatch: any): AdminApprovalsMappedActions => ({
  fetchRevisionDetail: param => dispatch(revisionDetailRequestData(param)),
  fetchAppDetail: (id: string) => dispatch(appDetailRequestData({ id }))
})

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(AdminApprovals)
)
