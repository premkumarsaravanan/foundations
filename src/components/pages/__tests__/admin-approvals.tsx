import * as React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import {
  AdminApprovals,
  AdminApprovalsProps,
  mapStateToProps,
  mapDispatchToProps,
  handleAfterClose,
  handleOnPageChange
} from '../admin-approvals'
import { approvalsStub } from '@/sagas/__stubs__/approvals'
import { AdminApprovalsList } from '@/reducers/admin-approvals'
import { ReduxState } from '@/types/core'
import routes from '@/constants/routes'

const mockProps = (loading: boolean, approvals: AdminApprovalsList | null): AdminApprovalsProps => ({
  approvalsState: {
    loading: loading,
    adminApprovalsData: approvals
  },
  // @ts-ignore: just pick the needed props for the test
  match: {
    params: {
      page: '2'
    }
  }
})

describe('AdminApproval', () => {
  it('should match a snapshot when LOADING false', () => {
    expect(toJson(shallow(<AdminApprovals {...mockProps(false, approvalsStub)} />))).toMatchSnapshot()
  })

  it('should match a snapshot when LOADING true', () => {
    expect(toJson(shallow(<AdminApprovals {...mockProps(true, null)} />))).toMatchSnapshot()
  })

  describe('mapStateToProps', () => {
    it('should run correctly', () => {
      const mockState = {
        adminApprovals: {},
        appDetail: {},
        revisionDetail: {}
      } as ReduxState
      const output = {
        approvalsState: mockState.adminApprovals,
        appDetail: mockState.appDetail,
        revisionDetail: mockState.revisionDetail
      }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(output)
    })
  })

  describe('mapDispatchToProps', () => {
    it('fetchRevisionDetail', () => {
      const mockDispatch = jest.fn()
      const { fetchRevisionDetail } = mapDispatchToProps(mockDispatch)
      fetchRevisionDetail({ appId: '123', appRevisionId: '123' })
      expect(mockDispatch).toBeCalled()
    })

    it('fetchAppDetail', () => {
      const mockDispatch = jest.fn()
      const { fetchAppDetail } = mapDispatchToProps(mockDispatch)
      fetchAppDetail('123')
      expect(mockDispatch).toBeCalled()
    })
  })

  describe('handleAfterClose', () => {
    it('should call dispatch', () => {
      const setIsModalOpen = jest.fn()
      const fn = handleAfterClose({ setIsModalOpen })
      fn()
      expect(setIsModalOpen).toBeCalled()
    })
  })

  describe('handleOnPageChange', () => {
    it('should call dispatch', () => {
      const mockHistory = {
        push: jest.fn()
      }
      const fn = handleOnPageChange(mockHistory)
      fn('1')
      expect(mockHistory.push).toBeCalledWith(`${routes.ADMIN_APPROVALS}/${1}`)
    })
  })
})
