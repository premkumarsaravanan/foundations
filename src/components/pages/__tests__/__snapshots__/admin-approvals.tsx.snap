// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminApproval should match a snapshot when LOADING false 1`] = `
<Connect(ErrorBoundary)>
  <div
    className="container isRelative py-8"
    data-test="revision-list-container"
  >
    <div
      className="pin absolute flex items-center justify-center"
    >
      <Loader />
    </div>
    <table
      className="table isFullwidth"
      data-test="revision-list"
    >
      <thead>
        <tr>
          <th>
            #
          </th>
          <th>
            AppId
          </th>
          <th>
            Type
          </th>
          <th>
            Description
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr
          data-app-id="ddc8b25e-42a2-468e-aa8f-4091eee6958f"
          key="4afb0bb9-c1f8-403a-8fd6-1c92b4244906"
        >
          <th>
            1
          </th>
          <th>
            ddc8b25e-42a2-468e-aa8f-4091eee6958f
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_ddc8b25e-42a2-468e-aa8f-4091eee6958f"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="c9cbb8cd-74c3-4cc7-b632-d8e86b0b7d26"
          key="0d55cee6-b8c0-4fb9-be45-506c5992d267"
        >
          <th>
            2
          </th>
          <th>
            c9cbb8cd-74c3-4cc7-b632-d8e86b0b7d26
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_c9cbb8cd-74c3-4cc7-b632-d8e86b0b7d26"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="bbd8ed62-64a5-4bb9-a269-7aa987923897"
          key="1c6f4453-06c6-424b-855a-9a77a718f83f"
        >
          <th>
            3
          </th>
          <th>
            bbd8ed62-64a5-4bb9-a269-7aa987923897
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_bbd8ed62-64a5-4bb9-a269-7aa987923897"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="38c30c86-984f-4e20-8133-93c42f69a283"
          key="371e1640-6d07-413c-8be2-214a6ee32c14"
        >
          <th>
            4
          </th>
          <th>
            38c30c86-984f-4e20-8133-93c42f69a283
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_38c30c86-984f-4e20-8133-93c42f69a283"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="b50fe8e3-0195-4ce5-9150-d6b6250aa93f"
          key="ca0e95f2-8dbe-4d62-b533-a7c695f5de9b"
        >
          <th>
            5
          </th>
          <th>
            b50fe8e3-0195-4ce5-9150-d6b6250aa93f
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_b50fe8e3-0195-4ce5-9150-d6b6250aa93f"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="ce069a76-6be1-4640-b412-b3c2a1e80dff"
          key="9ec2b8e8-209c-422f-a221-7d45e832a848"
        >
          <th>
            6
          </th>
          <th>
            ce069a76-6be1-4640-b412-b3c2a1e80dff
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_ce069a76-6be1-4640-b412-b3c2a1e80dff"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="e5854bf2-8a2b-4d41-b51b-c226dbac3ffb"
          key="c5d25de8-93e7-4f64-af41-e35833868ec4"
        >
          <th>
            7
          </th>
          <th>
            e5854bf2-8a2b-4d41-b51b-c226dbac3ffb
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_e5854bf2-8a2b-4d41-b51b-c226dbac3ffb"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="2421665e-7154-4bf7-bf7b-ad80bc7ed191"
          key="ff820e01-6444-409f-be13-466b6f527c76"
        >
          <th>
            8
          </th>
          <th>
            2421665e-7154-4bf7-bf7b-ad80bc7ed191
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_2421665e-7154-4bf7-bf7b-ad80bc7ed191"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="2baf738b-ab70-4de2-8a62-547a5ff68866"
          key="76b5b475-22d9-4437-b9bf-44a6afb5df14"
        >
          <th>
            9
          </th>
          <th>
            2baf738b-ab70-4de2-8a62-547a5ff68866
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_2baf738b-ab70-4de2-8a62-547a5ff68866"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
        <tr
          data-app-id="d36137d1-cc86-4b9a-a164-069183065f37"
          key="c6d3160b-b299-4764-87bf-e1f26065d282"
        >
          <th>
            10
          </th>
          <th>
            d36137d1-cc86-4b9a-a164-069183065f37
          </th>
          <th>
            App Revision
          </th>
          <th>
            A developer has submitted a request to modify their application
          </th>
          <th>
            <button
              className="button isPrimary"
              data-test="view-details-button_d36137d1-cc86-4b9a-a164-069183065f37"
              onClick={[Function]}
            >
              View details
            </button>
          </th>
        </tr>
      </tbody>
    </table>
    <Pagination
      onChange={[Function]}
      pageNumber={2}
      pageSize={10}
      totalCount={69}
    />
  </div>
  <Component
    afterClose={[Function]}
    visible={false}
  />
</Connect(ErrorBoundary)>
`;

exports[`AdminApproval should match a snapshot when LOADING true 1`] = `<Loader />`;
